internal/spice/config.go

```diff
@@ -0,0 +1,143 @@
+package spice
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/alecthomas/kong"
+	"github.com/charmbracelet/log"
+	"go.abhg.dev/gs/internal/git"
+)
+
+const (
+	_configKey             = "config"
+	_configNamespace       = "spice"
+	_configNamespacePrefix = _configNamespace + "."
+)
+
+// GitConfigLister provides access to git-config output.
+type GitConfigLister interface {
+	ListRegexp(context.Context, string) (
+		func(yield func(git.ConfigEntry, error) bool),
+		error,
+	)
+}
+
+var _ GitConfigLister = (*git.Config)(nil)
+
+// Config defines the git-spice configuration source.
+// It can be passed into Kong as a [kong.Resolver] to fill in flag values.
+//
+// Configuration for git-spice is specified via git-config.
+// These can be system, user, repository, or worktree-level.
+//
+// The configuration keys are read from the root namespace "spice"
+// for keys in the CLI grammar tagged with the `config:"key"` tag.
+// Flags that are not tagged with `config:"key"` are ignored for configuration.
+//
+// For example:
+//
+//	type someCmd struct {
+//		Level string `config:"level"`
+//	}
+//
+// This will read the configuration key "spice.level" from git-config.
+//
+//	[spice]
+//	level = hot
+//
+// Values are decoded using the same mapper as the flag.
+// For single-valued fields, if multiple values are found in the configuration,
+// the last value is used.
+// For slice fields, all values are combined.
+//
+// If a flag is passed on the CLI, it takes precedence over the configuration.
+type Config struct {
+	// items is a map from configuration key (without the "spice." prefix)
+	// to list of values for that field.
+	items map[string][]string
+}
+
+// ConfigOptions spceifies options for the [Config].
+type ConfigOptions struct {
+	// Log specifies the logger to use for logging.
+	// Defaults to no logging.
+	Log *log.Logger
+}
+
+// LoadConfig loads configuration from the provided [GitConfig].
+func LoadConfig(ctx context.Context, cfg GitConfigLister, opts ConfigOptions) (*Config, error) {
+	if opts.Log == nil {
+		opts.Log = log.New(io.Discard)
+	}
+
+	entries, err := cfg.ListRegexp(ctx, `^`+_configNamespace+`\.`)
+	if err != nil {
+		return nil, fmt.Errorf("list configuration: %w", err)
+	}
+
+	items := make(map[string][]string)
+
+	err = nil // TODO: use a range loop after Go 1.23
```

Comment:
You used return on err not nil, so I don't get this


```suggestion
	// TODO: use a range loop after Go 1.23
```

Reply:
So the `err` is being filled in from the iteration loop below:

```
err = nil
entries(func(..., fooErr error) bool {
  if fooErr != nil {
    err = fooErr
    return false
  }
  // ...
})
```

Once Go 1.23 is released, this will all become:

```
for entry, err := range entries {
  if err != nil {
    return fmt.Errorf("read configuration: %w", err)
  }
  // ...
}
```

Reply:
Oh, that's right. Thanks for the explanation